062-quotas-management.md

```diff
@@ -0,0 +1,163 @@
+   Having a dedicated API for quota configuration would make it much easier to have the operator configure the client automatically.
+
+## Proposal
+
+This proposal suggests to add a new `quotas` API to the `Kafka` custom resource.
+It will be inside the `.spec.kafka` section.
+And it will allow configuring different types of quotas.
+As many existing Strimzi APIs, it will use the `type` field to differentiate the quota plugin that will be used.
+
+This proposal suggests to initially add support for two Quota plugins.
+1. The build-in Kafka plugin
+2. The Strimzi Quotas plugin
+
+The first supported plugin - the Kafka's built-in plugin - will be configured as `type: kafka`.
+The API will allow to configure the default user quotas for this plugin in the same format as the `KafkaUser` API allows it today:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: kafka
+      producerByteRate: 1048576
+      consumerByteRate: 2097152
+      requestPercentage: 55
+      controllerMutationRate: 50
+    # ... 
+  # ...
+```
+
+Based on this configuration, the operator will:
+* Configure the default quota plugin (make sure the `client.quota.callback.class` option is not set)
+* Set the default user quotas
+* When the `type: kafka` quotas APi will be set, but one or more of fields with the individual quotas will be missing, the operator will make sure that given default quota is not set.
+* The other configuration options for the built-in Kafka plugin can be configured by the user in the `.spec.kafka.config` section in the `Kafka` CR.
+
+The second supported plugin - the Strimzi Quotas plugin - will be configured as `type: strimzi`.
+The API will allow to configure the storage limits at which the plugin will start limiting the producers in order to prevent the disk from getting full.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    quotas:
+      type: strimzi
+      produceQuota: 1000000
+      fetchQuota: 1000000
+      minAvailableBytesPerVolume: 5368709120
+      excludedPrincipals:
+        - my-special-user
+    # ... 
+  # ...
+```
+
+The `type: strimzi` quotas API will support the following fields:
+* `produceQuota` for setting the default produce quota
+* `fetchQuota` for setting the default fetch quota
+* `minAvailableBytesPerVolume` for setting the minimal required free space per volume as number of bytes (this field can be configured only when `minAvailableRatioPerVolume`is not set)
+* `minAvailableRatioPerVolume` for setting the minimal required free space per volume as percentage of the overall volume size (this field can be configured only when `minAvailableBytesPerVolume`is not set)
+* `excludedPrincipals` for setting s list of users to be excluded from the applied quotas
```

Comment:
Obviously this is equivalent to what the proposed version of the quota plug-in offers, but shouldn't it be driven by the plug-in rather than explicitly by this (or some other subsequent proposal)? Otherwise how would a user pass their own custom configuration through to their chosen plug-in?

Reply:
I don't follow this. Strimzi controls the configuration. Something like _own custom configuration_ is something what is really available only in _seelcted places_ because there was no way how to reasonably validate it. But this does not seem to be the case.

Reply:
I suspect I'm over thinking it, but I was imagining users being able to configure their own quota plug-ins not just the Kafka and Strimzi provided ones, but they can still do that directly without operator support so we don't need to support arbitrary config blobs here.

Reply:
We could in theory add some third type `type: custom` where users can freely configure everything. But as we have to let them also do it directly in `.spec.kafka.config`, I'm not sure it would have any value since they can do it anyway there. I guess it would make sense only if we bundle this inside a feature gate to somehow disable the quotas configuration in `.spec.kafka.config` after some time. But I'm not sure that is really useful and easily implementable.